File: back/models/event.go
package models

import (
	"gorm.io/gorm"
)

const DateFormat = "2006-01-02"
const TimeFormat = "15:04"

type RecurrenceType string

const (
	EachDays   RecurrenceType = "eachDays"
	EachWeeks  RecurrenceType = "eachWeeks"
	EachMonths RecurrenceType = "eachMonths"
	EachYears  RecurrenceType = "eachYears"
)

type Event struct {
	gorm.Model
	Name           string          `json:"name" validate:"required,min=2,max=100"`
	Description    string          `json:"description" validate:"required"`
	Date           string          `json:"date" validate:"required,datetime=2006-01-02"`
	Time           *string         `json:"time" validate:"omitempty,datetime=15:04"`
	TypeID         uint            `json:"type_id"`
	Type           *EventType      `json:"type,omitempty"`
	AddressID      uint            `json:"address_id"`
	Address        *Address        `json:"address,omitempty"`
	OrganizerID    uint            `json:"organizer_id"`
	Organizer      *User           `json:"organizer,omitempty"`
	GroupID        uint            `json:"group_id"`
	Group          *Group          `json:"group,omitempty"`
	Participants   []User          `gorm:"many2many:attends;" json:"participants,omitempty"`
	RecurrenceType *RecurrenceType `json:"recurrence_type,omitempty" validate:"omitempty,oneof=eachDays eachWeeks eachMonths eachYears"`
}

type EventCreate struct {
	Name           string          `json:"name" validate:"required,min=2,max=100"`
	Description    string          `json:"description" validate:"required"`
	Date           string          `json:"date" validate:"required,datetime=2006-01-02"`
	Time           *string         `json:"time" validate:"omitempty,datetime=15:04"`
	TypeID         uint            `json:"type_id" validate:"required"`
	GroupID        uint            `json:"group_id" validate:"required"`
	OrganizerID    uint            `json:"-" validate:"required"`
	Address        *AddressCreate  `json:"address" validate:"required"`
	RecurrenceType *RecurrenceType `json:"recurrence_type,omitempty" validate:"omitempty,oneof=eachDays eachWeeks eachMonths eachYears"`
}

func (e EventCreate) ToEvent() *Event {
	return &Event{
		Name:           e.Name,
		Description:    e.Description,
		Date:           e.Date,
		Time:           e.Time,
		TypeID:         e.TypeID,
		GroupID:        e.GroupID,
		Address:        e.Address.ToAddress(),
		OrganizerID:    e.OrganizerID,
		RecurrenceType: e.RecurrenceType,
	}
}

File: front/lib/core/models/event.dart
import "package:front/core/models/event_type.dart";
import "package:front/core/models/user.dart";

import "address.dart";

enum RecurrenceType {
  eachDays,
  eachWeeks,
  eachMonths,
  eachYears,
}

class Event {
  final int id;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deletedAt;
  final String name;
  final String description;
  final DateTime date;
  final String? time;
  final int typeId;
  final EventType? type;
  final int organizerId;
  final int addressId;
  final Address? address;
  final User? organizer;
  final int groupId;
  final RecurrenceType? recurrenceType;

  Event({
    required this.id,
    required this.createdAt,
    required this.updatedAt,
    this.deletedAt,
    required this.name,
    required this.description,
    required this.date,
    this.time,
    required this.typeId,
    this.type,
    required this.organizerId,
    required this.addressId,
    this.address,
    this.organizer,
    required this.groupId,
    this.recurrenceType,
  });

  factory Event.fromJson(Map<String, dynamic> json) {
    return Event(
      id: json['ID'],
      createdAt: DateTime.parse(json['CreatedAt']),
      updatedAt: DateTime.parse(json['UpdatedAt']),
      deletedAt:
          json['DeletedAt'] != null ? DateTime.parse(json['deleted_at']) : null,
      name: json['name'],
      description: json['description'],
      date: DateTime.parse(json['date']),
      time: json['time'],
      typeId: json['type_id'],
      organizerId: json['organizer_id'],
      addressId: json['address_id'],
      address:
          json['address'] != null ? Address.fromJson(json['address']) : null,
      organizer:
          json['organizer'] != null ? User.fromJson(json['organizer']) : null,
      type: json['type'] != null ? EventType.fromJson(json['type']) : null,
      groupId: json['group_id'],
      recurrenceType: json['recurrence_type'] != null
          ? RecurrenceType.values.firstWhere((e) =>
              e.toString() == 'RecurrenceType.' + json['recurrence_type'])
          : null,
    );
  }
}

class EventCreate {
  String name;
  String description;
  String date;
  String time;
  int typeId;
  int groupId;
  AddressCreate address;
  RecurrenceType? recurrenceType;

  EventCreate({
    required this.name,
    required this.description,
    required this.date,
    required this.time,
    required this.typeId,
    required this.groupId,
    required this.address,
    this.recurrenceType,
  });

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'description': description,
      'date': date,
      'time': time,
      'type_id': typeId,
      'group_id': groupId,
      'address': address.toJson(),
      'recurrence_type': recurrenceType?.toString().split('.').last,
    };
  }
}
